### PR TITLE
[RLlib] Fix bugs in `MultiAgentEpisode` class (getters, `env_t_to_agent_t` system, etc..).

### PR DESCRIPTION
**Pull request by sven1977**
Fix bugs in `MultiAgentEpisode` class:
* Fix `env_t_to_agent_t` system.
* Cleanup some unneeded code.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/env/multi_agent_episode.py [hunk 1]
```diff
@@ -941,15 +937,15 @@ def cut(self, len_lookback_buffer: int = 0) -> "MultiAgentEpisode":
...
+        orig_len_lb = len_lookback_buffer
         for agent_id, agent_actions in self._hanging_actions_end.items():
             assert self.env_t_to_agent_t[agent_id].get(-1) == self.SKIP_ENV_TS_TAG
-            for i in range(1, self.env_t_to_agent_t[agent_id].lookback + 1):
-                if (
-                    self.env_t_to_agent_t[agent_id].get(-i, neg_index_as_lookback=True)
-                    != self.SKIP_ENV_TS_TAG
-                ):
-                    len_lookback_buffer = max(len_lookback_buffer, i)
+            for i in range(orig_len_lb, len(self.env_t_to_agent_t[agent_id].data) + 1):
```
Comment (simonsays1980) [inline rllib/env/multi_agent_episode.py:945]: Why do we access here directly the raw `data` attribute when we have the `InfiniteLookbackBuffer` defined as a class with certain length methods?

File: rllib/env/multi_agent_episode.py [hunk 2]
```diff
@@ -2325,6 +2293,8 @@ def _get_data_by_env_steps_as_list(
                     hanging_val,
                     filter_for_skip_indices=idxes[i],
                 )
+                if what == "extra_model_outputs" and not inf_lookback_buffer:
```
Comment (simonsays1980) [inline rllib/env/multi_agent_episode.py:2296]: I guess `inf_lookback_buffer` is a bool. Imo the code becomes better readable, if we change the return values of `_get_inf_lookback_buffer_or_dict` to a `indices_to_use` only and instead check here for the type.

File: rllib/env/multi_agent_episode.py [hunk 3]
```diff
@@ -2532,12 +2502,15 @@ def _get_single_agent_data_by_index(
...
                 # buffers.
                 if extra_model_outputs_key is None:
+                    assert hanging_val is None or isinstance(hanging_val, dict)
                     return {
                         key: sub_buffer.get(
                             indices=index_incl_lookback - sub_buffer.lookback,
                             neg_index_as_lookback=True,
                             fill=fill,
-                            _add_last_ts_value=hanging_val,
+                            _add_last_ts_value=(
```
Comment (simonsays1980) [inline rllib/env/multi_agent_episode.py:2511]: Tricky. How did you find it?

File: rllib/env/tests/test_multi_agent_episode.py [hunk 4]
```diff
@@ -2254,6 +2254,31 @@ def test_other_getters(self):
...
+            [
+                {"a0": 0},
+                {"a2": 0},
+                {"a2": 1},
+                {"a2": 2},
+                {"a0": 1},
+                {"a2": 3},
+                {"a2": 4},
+
+                # <- BUT: actual cut here, b/c of hanging action of a2
```
Comment (simonsays1980) [inline rllib/env/tests/test_multi_agent_episode.py:2269]: Can we clarify here why this action is hanging? And means "cut" `MAE.cut`?

File: rllib/env/utils/infinite_lookback_buffer.py [hunk 5]
```diff
@@ -533,9 +533,18 @@ def _get_int_index(
...
         data_to_use = self.data
         if _ignore_last_ts:
-            data_to_use = self.data[:-1]
+            if self.finalized:
+                data_to_use = tree.map_structure(lambda s: s[:-1], self.data)
+            else:
+                data_to_use = self.data[:-1]
         if _add_last_ts_value is not None:
-            data_to_use = np.append(data_to_use.copy(), _add_last_ts_value)
+            if self.finalized:
```
Comment (simonsays1980) [inline rllib/env/utils/infinite_lookback_buffer.py:541]: Yeah, I remember this location to be part of the error output.

**Review by simonsays1980 — APPROVED** [review]
LGTM. Some small nits in regard to commenting.

File: rllib/env/multi_agent_episode.py (hunk 1 above)
Comment (sven1977) [reply rllib/env/multi_agent_episode.py:945]: Great question and yes, this is a little hacky here. The len method returns the length but w/o(!) the lookback buffer.

File: rllib/env/tests/test_multi_agent_episode.py (hunk 4 above)
Comment (sven1977) [reply rllib/env/tests/test_multi_agent_episode.py]: 👍  done.

File: rllib/env/utils/infinite_lookback_buffer.py (hunk 5 above)
Comment (sven1977) [reply rllib/env/utils/infinite_lookback_buffer.py:541]: Perfect! Thanks for confirming! This was a pretty common bug you would run into when your multi-agent episode is a little bit more unstructured wrt agents stepping patterns.

**Pull request merged**